app/src/main/java/com/github/se/cyrcle/model/network/ConnectivityObserver.kt

```diff
@@ -0,0 +1,80 @@
+package com.github.se.cyrcle.model.network
+
+import android.content.Context
+import android.util.Log
+import com.github.se.cyrcle.model.map.MapViewModel
+import com.github.se.cyrcle.model.parking.ParkingViewModel
+import com.github.se.cyrcle.model.user.UserViewModel
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+class ConnectivityObserver(
+    private val userViewModel: UserViewModel,
+    private val mapViewModel: MapViewModel,
+    private val parkingViewModel: ParkingViewModel
+) {
+  private lateinit var networkReceiver: NetworkReceiver
+  /** Initializes the Connectivity observer. */
+  fun init(context: Context) {
+    initNetworkReceiver(context)
+    initUserConnectionModeObserver()
+  }
+  /**
+   * Initializes the network receiver and registers it It Listen for network changes and update the
+   * state accordingly
+   */
+  private fun initNetworkReceiver(context: Context) {
+    networkReceiver = NetworkReceiver(context)
+    networkReceiver.register()
+    CoroutineScope(Dispatchers.Default).launch {
+      networkReceiver.isConnected.collect {
+        userViewModel.setHasConnection(it)
+        Log.d("ConnectivityObserver", "User has connection : ${networkReceiver.isConnected.value}")
+      }
+    }
+  }
+
+  /**
+   * Initializes the user connection mode observer This will listen for changes in the user
+   * connection mode and update the viewmodels accordingly
+   */
+  private fun initUserConnectionModeObserver() {
+    CoroutineScope(Dispatchers.Default).launch {
+      userViewModel.isOnlineMode.collect { isOnlineMode ->
+        if (isOnlineMode) onOnlineMode() else onOfflineMode()
+      }
+    }
+  }
+
+  /**
+   * Switches the app to online mode by
+   * - signing in the user anonymously
+   * - switching the map and parking viewmodels to online mode (Soon)
```

Comment:
The coaches recommend to add TODO. (im not making the rules here)

Reply:
Done

Reply:
this is not a TODO... a todo is in flashy yellow (therefore why it's recommanded) and you can write one with : 
//TODO "your_text"
And my remark also regards the function below which also needs it.